### PR TITLE
Agilent33500 multichannel

### PR DESCRIPTION
**Pull request by JCarl-OS**
Adding channel feature to Agilent33500 series instrument

### CODE REVIEW

File: pymeasure/instruments/agilent/agilent33500.py [hunk 1]
```diff
@@ -56,58 +365,58 @@ class Agilent33500(Instrument):
...
         generator = Agilent33500("GPIB::1")
 
-        generator.shape = 'SIN'                 # Sets the output signal shape to sine
-        generator.frequency = 1e3               # Sets the output frequency to 1 kHz
-        generator.amplitude = 5                 # Sets the output amplitude to 5 Vpp
-        generator.output = 'on'                 # Enables the output
+        generator.ch[n].shape = 'SIN'                 # Sets the output signal shape to sine
+        generator.ch[n].frequency = 1e3               # Sets the output frequency to 1 kHz
+        generator.ch[n].amplitude = 5                 # Sets the output amplitude to 5 Vpp
+        generator.ch[n].output = 'on'                 # Enables the output
```
Comment (msmttchr) [inline pymeasure/instruments/agilent/agilent33500.py:371]: I would replace `n` with an actual channel number to make the code working as an example.
Same comment in other example code below.

File: pymeasure/instruments/agilent/agilent33500.py [hunk 2]
```diff
@@ -48,6 +46,317 @@
 # string_validator = joined_validators(capitalize_string, strict_discrete_set)
 
 
+class Agilent33500Channel(Channel):
+    """Implementation of a base Agilent 33500 channel"""
+
+    def __init__(self, instrument, id):
```
Comment (msmttchr) [inline pymeasure/instruments/agilent/agilent33500.py:52]: I don't think you need to redefine `__init__`. Right?

File: pymeasure/instruments/agilent/agilent33500.py [hunk 3]
```diff
@@ -48,6 +46,317 @@
...
+        validator=strict_discrete_set,
+        values=["SIN", "SQU", "TRI", "RAMP", "PULS", "PRBS", "NOIS", "ARB", "DC"],
+    )
+
+    frequency = Instrument.control(
+        "SOUR{ch}:FREQ?",
+        "SOUR{ch}:FREQ %f",
+        """ A floating point property that controls the frequency of the output
+        waveform in Hz, from 1 uHz to 120 MHz (maximum range, can be lower depending
+        on your device), depending on the specified function. Can be set. """,
```
Comment (msmttchr) [inline pymeasure/instruments/agilent/agilent33500.py:69]:
```suggestion
        on your device), depending on the specified function.""",
```
I think "can be set" is implied by the verb control.
Similar comments for other places.

File: pymeasure/instruments/agilent/agilent33500.py [hunk 4]
```diff
@@ -48,6 +46,317 @@
...
+
+                            format = 'binary': Accepts a binary stream of 8 bit data.
+        :param data_format: Defines the format of data_points. Can be 'DAC' (default), 'float' or
+                            'binary'. See documentation on parameter data_points above.
+        """
+        if data_format == "DAC":
+            separator = ", "
+            data_points_str = [str(item) for item in data_points]  # Turn list entries into strings
+            data_string = separator.join(data_points_str)  # Join strings with separator
+            print(f"SOUR{{ch}}:DATA:ARB:DAC {arb_name}, {data_string}")
```
Comment (msmttchr) [inline pymeasure/instruments/agilent/agilent33500.py:340]: I guess this leftover from debug code. Right?

File: pymeasure/instruments/agilent/agilent33500.py [hunk 5]
```diff
@@ -48,6 +46,317 @@
...
+        """ An floating point property that sets the sample rate of the currently selected
+        arbitrary signal. Valid values are 1 µSa/s to 250 MSa/s (maximum range, can be lower
+        depending on your device). This can be set. """,
+        validator=strict_range,
+        values=[1e-6, 250e6],
+    )
+
+    def data_volatile_clear(self):
+        """
+        Clear all arbitrary signals from the volatile memory. This should be done if the same name
```
Comment (msmttchr) [inline pymeasure/instruments/agilent/agilent33500.py:305]: I think docstring should have a single line description and
 a more detailed description after a blank line.
Same comment for other docstring.

File: pymeasure/instruments/agilent/agilent33500.py [hunk 6]
```diff
@@ -226,16 +547,18 @@ def beep(self):
...
     output = Instrument.control(
-        "OUTP?", "OUTP %d",
+        "OUTP?",
+        "OUTP %d",
         """ A boolean property that turns on (True, 'on') or off (False, 'off')
         the output of the function generator. Can be set. """,
         validator=strict_discrete_set,
         map_values=True,
-        values={True: 1, 'on': 1, 'ON': 1, False: 0, 'off': 0, 'OFF': 0},
+        values={True: 1, "on": 1, "ON": 1, False: 0, "off": 0, "OFF": 0},
```
Comment (msmttchr) [inline pymeasure/instruments/agilent/agilent33500.py:556]: I see many changes involving replacement of single quote with double quote and viceversa. Are these changes really needed?

**Review by msmttchr — COMMENTED** [review]
First round of comments

File: pymeasure/instruments/agilent/agilent33500.py [hunk 7]
```diff
@@ -56,58 +365,58 @@ class Agilent33500(Instrument):
...
-        generator.arb_file = 'test'             # Select the transmitted waveform 'test'
+        generator.ch[n].arb_file = 'test'             # Select the transmitted waveform 'test'
 
     """
 
-    id = Instrument.measurement(
-        "*IDN?", """ Reads the instrument identification """
-    )
+    ch = Instrument.ChannelCreator(Agilent33500Channel, (1, 2))
+    id = Instrument.measurement("*IDN?", """ Reads the instrument identification """)
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: That is correct. Removed, and tested. 👍

File: pymeasure/instruments/agilent/agilent33500.py (hunk 1 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Modified in https://github.com/pymeasure/pymeasure/pull/763/commits/3e71c1eee00e7ff48596b6418855cf6a67fe831e 
I added an extra comment about the default channel

File: pymeasure/instruments/agilent/agilent33500.py (hunk 2 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: You are correct. Removed in and tested in https://github.com/pymeasure/pymeasure/pull/763/commits/3e71c1eee00e7ff48596b6418855cf6a67fe831e

File: pymeasure/instruments/agilent/agilent33500.py (hunk 3 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Fair. Removed a bunch of "can be set" in https://github.com/pymeasure/pymeasure/pull/763/commits/3e71c1eee00e7ff48596b6418855cf6a67fe831e

File: pymeasure/instruments/agilent/agilent33500.py (hunk 4 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: It seems to be from previous commits, but you are right, when tested with pytest -s it threw tons of data prints. Removed in https://github.com/pymeasure/pymeasure/pull/763/commits/3e71c1eee00e7ff48596b6418855cf6a67fe831e

File: pymeasure/instruments/agilent/agilent33500.py (hunk 5 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Re-formatted all dosctrings with that issue (I hope) in https://github.com/pymeasure/pymeasure/pull/763/commits/3e71c1eee00e7ff48596b6418855cf6a67fe831e

File: pymeasure/instruments/agilent/agilent33500.py (hunk 6 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py:560]: Before committing I reformated the file with black. The replacements are not needed as in they don't break anything either way. While black is PEP 8 compliant, it is true it has its own style. Let me know what style the project prefers and I can change the device and tests to it.

Comment (bmoneke) [reply pymeasure/instruments/agilent/agilent33500.py:560]: The reformatting (also splitting/merging lines, changing number formatting) makes it hard to see the actual implementation details.
In fact, we have black style configuration in this project (I personally do not use black, but it was recently added).

Here a quote from our [documentation](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html?highlight=black#python-style-guides), how to use it (for better code review):

> "It is allowed but not required to use the [black](https://black.readthedocs.io/en/stable/) code formatter. Do only format files that you are actively working on and make a dedicated commit for the initial formatting when working on an existing file that does not adhere to the black code style to facilitate code review. If there are conflicts between black’s output and flake8 (especially related to [E203](https://www.flake8rules.com/rules/E203.html)), flake8 takes precedence. Use #noqa : E203 to disable E203 warnings for a specific line if appropriate."

**Review by bmoneke — CHANGES_REQUESTED** [review]
Thank you @JCarl-OS for your contribution. It is good to have people contributing new instruments and fixing parts.
You may add yourself to `authors.txt` as well.

I have a few general comments besides the comments I made in the code:

- We recently changed the way to write [docstrings](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html?#docstrings) (see examples in [adding instruments](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#adding-instruments)) of `control` etc. (I saw you removed "can be set." in quite a few). Please be so kind and change the docstrings of the properties **you added / changed significantly**.
- I made a comment regarding the use of black, see that and make that, if possible.
- It is good if you add protocol tests (which do not require a connected instrument), for details see the comment in your test file.

**Comment by msmttchr** [comment]
@JCarl-OS in order to anticipate link failures, you could install locally `flake8 `and run in your local copy:
`flake8 . --count --extend-select=E9,F63,F7,F82 --show-source --statistics`.
This will allow to save a lot of time compared to check errors after CI actions.

**Comment by bilderbuchi** [comment]
> @JCarl-OS in order to anticipate link failures, you could install locally `flake8 `and run in your local copy: `flake8 . --count --extend-select=E9,F63,F7,F82 --show-source --statistics`. This will allow to save a lot of time compared to check errors after CI actions.

Actually, just `flake8 .` should be sufficient, that extra fluff is just so that it's a bit easier to notice stuff in the CI logs.

File: pymeasure/instruments/agilent/agilent33500.py (hunk 6 above)
Comment (bilderbuchi) [reply pymeasure/instruments/agilent/agilent33500.py:560]: I proposed (#767) to use https://github.com/akaihola/darker instead of black to limit the fallout from the formatting -- could you maybe retry with that tool, instead, and report back your experience?

File: pymeasure/instruments/agilent/agilent33500.py [hunk 8]
```diff
@@ -48,138 +46,461 @@
...
+        validator=strict_range,
+        values=[-4.995, +4.995],
+    )
+
+    voltage_high = Instrument.control(
+        "SOUR{ch}:VOLT:HIGH?",
+        "SOUR{ch}:VOLT:HIGH %f",
+        """ A floating point property that controls the upper voltage of the
+        output waveform in V, from -4.990 V to 5 V (must be higher than low
+        voltage by at least 1 mV).""",
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py:106]: Nope, you're right. I can test and modify accordingly

File: tests/instruments/agilent/test_agilent33500_with_device.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import pytest
+from pymeasure.instruments.agilent.agilent33500 import Agilent33500
+from math import pi, sin
+
+
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply tests/instruments/agilent/test_agilent33500_with_device.py:29]: Added skip in https://github.com/pymeasure/pymeasure/pull/763/commits/88bee61449ed1016ebe027a31a365be15a327414

File: tests/instruments/agilent/test_agilent33500_with_device.py [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+# from pyvisa.errors import VisaIOError
+
+############
+# FIXTURES #
+############
+
+
+@pytest.fixture(scope="session")
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply tests/instruments/agilent/test_agilent33500_with_device.py:36]: I would not mind adding protocol tests at all. I should be able to work on them in the next few days

File: pymeasure/instruments/agilent/agilent33500.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Figured out the instrument can do -4.999. Updated ranges and docstrings in https://github.com/pymeasure/pymeasure/pull/763/commits/88bee61449ed1016ebe027a31a365be15a327414

File: pymeasure/instruments/agilent/agilent33500.py [hunk 11]
```diff
@@ -48,138 +46,461 @@
...
+        for the ramp waveform, from 0.0% to 100.0%.""",
+        validator=strict_range,
+        values=[0, 100],
+    )
+
+    pulse_period = Instrument.control(
+        "SOUR{ch}:FUNC:PULS:PER?",
+        "SOUR{ch}:FUNC:PULS:PER %e",
+        """ A floating point property that controls the period of a pulse
+        waveform function in seconds, ranging from 33 ns to 1e6 s. Can be set
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Addressed in https://github.com/pymeasure/pymeasure/pull/763/commits/88bee61449ed1016ebe027a31a365be15a327414

File: pymeasure/instruments/agilent/agilent33500.py [hunk 12]
```diff
@@ -48,138 +46,461 @@
...
+        validator=strict_range,
+        values=range(1, 100000),
+    )
+
+    arb_file = Instrument.control(
+        "SOUR{ch}:FUNC:ARB?",
+        "SOUR{ch}:FUNC:ARB %s",
+        """ A string property that selects the arbitrary signal from the volatile
+        memory of the device. String has to match an existing arb signal in volatile
+        memore (set by data_arb()).""",
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Followed suggestion in https://github.com/pymeasure/pymeasure/pull/763/commits/88bee61449ed1016ebe027a31a365be15a327414

File: pymeasure/instruments/agilent/agilent33500.py [hunk 13]
```diff
@@ -48,138 +46,461 @@
...
+
+    def data_arb(self, arb_name, data_points, data_format="DAC"):
+        """
+        Uploads an arbitrary trace into the volatile memory of the device for a given channel.
+
+        The data_points can be given as comma separated 16 bit DAC values (ranging from -32767 to +32767),
+        as comma separated floating point values (ranging from -1.0 to +1.0) or as a binary data stream.
+        Check the manual for more information. The storage depends on the device type and ranges
+        from 8 Sa to 16 MSa (maximum).
+        TODO: *Binary is not yet implemented*
```
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: Thanks for the feedback. Moved in https://github.com/pymeasure/pymeasure/pull/763/commits/88bee61449ed1016ebe027a31a365be15a327414

File: pymeasure/instruments/agilent/agilent33500.py [hunk 14]
```diff
@@ -48,138 +46,462 @@
...
-        generator.arb_file = 'test'             # Select the transmitted waveform 'test'
+         )
+        generator.ch[1].arb_file = 'test'       # Select the transmitted waveform 'test'
 
     """
 
-    id = Instrument.measurement(
-        "*IDN?", """ Reads the instrument identification """
-    )
+    ch = Instrument.ChannelCreator(Agilent33500Channel, (1, 2))
```
Comment (bmoneke) [inline pymeasure/instruments/agilent/agilent33500.py:392]: Just for your information:
The ChannelCreator will add also the channels as `inst.ch_1` and `inst.ch_2` (being an alternative to `inst.ch[1]`).

File: tests/instruments/agilent/test_agilent33500_with_device.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/agilent/test_agilent33500_with_device.py:29]: You added the test just to that single function, because you _decorated_ that function with `@pytest...`.

Just place the code (without `@`) at the beginning of the file (as I already suggested). You may just hit "commit suggestion", and remove that other (then obsolete) line 77.

File: pymeasure/instruments/agilent/agilent33500.py (hunk 14 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py:392]: Thank you for the comment

File: pymeasure/instruments/agilent/agilent33500.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py]: @msmttchr Would you mind if I resolve the first round of comments?

File: pymeasure/instruments/agilent/agilent33500.py (hunk 6 above)
Comment (JCarl-OS) [reply pymeasure/instruments/agilent/agilent33500.py:561]: I guess I'm not clear on what the recommended action is. Please advise

**Comment by msmttchr** [comment]
Not at all, sorry for the delay. In any case, I have done it myself.
I have left open the one related to black, since we need to give feedback
on what to do.

Il Mer 30 Nov 2022, 04:58 Carlos Martinez ***@***.***> ha
scritto:

File: tests/instruments/agilent/test_agilent33500_with_device.py [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+PULSE_WIDTH_RANGE = [16e-9, 1e6]
+PULSE_DUTYCYCLE_RANGE = [0.1, 100]
+PULSE_TRANSITION_RANGE = [8.4e-9, 1e-6]
+BURST_MODES = [["TRIG", "TRIG"], ["TRIGGERED", "TRIG"], ["GAT", "GAT"], ["GATED", "GAT"]]
+BURST_PERIOD = [1e-6, 8000]
+BURST_NCYCLES = [1, 99999]
+SRATE = [1e-6, 1.6e8]
+
+
+@pytest.skip('Only works with connected hardware', allow_module_level=True)
```
Comment (bmoneke) [inline tests/instruments/agilent/test_agilent33500_with_device.py:79]: As you have it at module level, you do not need it here anymore. 
```suggestion
```

File: tests/instruments/agilent/test_agilent33500.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    ) as inst:
+        assert frequency == inst.ch[1].frequency
+        assert frequency == inst.ch[2].frequency
+        assert frequency == inst.frequency
+        inst.ch[1].frequency = frequency
+        inst.ch[2].frequency = frequency
+        inst.frequency = frequency
+
+
+@pytest.mark.xfail
```
Comment (bmoneke) [inline tests/instruments/agilent/test_agilent33500.py:84]: Why do you expect the tests to fail?

Please state the reason (in parenthesis) or make the test succeed.

File: tests/instruments/agilent/test_agilent33500.py [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+# THE SOFTWARE.
+#
+
+import pytest
+from pymeasure.test import expected_protocol
+from pymeasure.instruments.agilent.agilent33500 import Agilent33500
+
+
+# TODO: Tests with a range in floats, seems to be flaky. Need to
+#       figure out the reason for protocol returning 6 after decimal point
```
Comment (bmoneke) [inline tests/instruments/agilent/test_agilent33500.py:31]: I think 6 digits is the default precision of the formatter for floats.

You can use `%.2f` in the command string to change the precision to 2 digits.

**Review by bmoneke — APPROVED** [review]
Thanks for the protocol tests. I added a few remarks.

Besides these, the code looks good. Thanks.

File: pymeasure/instruments/agilent/agilent33500.py (hunk 6 above)
Comment (msmttchr) [reply pymeasure/instruments/agilent/agilent33500.py:561]: To me, you can leave the changes done by black.

File: tests/instruments/agilent/test_agilent33500.py (hunk 16 above)
Comment (JCarl-OS) [reply tests/instruments/agilent/test_agilent33500.py]: Removed xfail. I formatted the expected floats

File: tests/instruments/agilent/test_agilent33500.py (hunk 17 above)
Comment (JCarl-OS) [reply tests/instruments/agilent/test_agilent33500.py:31]: Done. Thanks for your input

File: tests/instruments/agilent/test_agilent33500.py [hunk 18]
```diff
@@ -68,9 +67,9 @@ def test_frequency(frequency):
             ("SOUR1:FREQ?", frequency),
             ("SOUR2:FREQ?", frequency),
             ("FREQ?", frequency),
-            (f"SOUR1:FREQ {frequency}", None),
-            (f"SOUR2:FREQ {frequency}", None),
-            (f"FREQ {frequency}", None),
+            (f"SOUR1:FREQ {'{:.6f}'.format(frequency)}", None),
+            (f"SOUR2:FREQ {'{:.6f}'.format(frequency)}", None),
+            (f"FREQ {'{:.6f}'.format(frequency)}", None),
```
Comment (bmoneke) [inline tests/instruments/agilent/test_agilent33500.py:72]: Why do you nest two format strings into each other?
The following should be sufficient as it does the same:
```suggestion
            (f"SOUR1:FREQ {frequency:.6f}", None),
            (f"SOUR2:FREQ {frequency:.6f}", None),
            (f"FREQ {frequency:.6f}", None),
```

In curly braces you can write the variable name (or even an expression) directly.

Comment (JCarl-OS) [reply tests/instruments/agilent/test_agilent33500.py]: Changed as requested in https://github.com/pymeasure/pymeasure/pull/763/commits/bab560446383b62f8ab8a32e97eea8ba9d8ea071

File: tests/instruments/agilent/test_agilent33500.py (hunk 17 above)
Comment (bmoneke) [reply tests/instruments/agilent/test_agilent33500.py:31]: As you understood, why it happens, as it happens, don't you want to remove those two lines?

**Comment by bmoneke** [comment]
@JCarl-OS are you ready to merge?

**Comment by JCarl-OS** [comment]
> @JCarl-OS are you ready to merge?

@bmoneke Yes
